File: crates/sc-core/src/aggregate.rs
//! Directory aggregate ETag —, and dirty marking on
//! the write path — §4.4. This is the one place `sc-core` allocates `sc-meta`
//! fileids for directories (files never need one just to be listed —
//! 's "lazy allocation"), because the aggregate cache is
//! keyed by fileid and computing it is the thing that actually needs that
//! identity to exist.
//!
//! Recursion depth here is bounded by `SafePath`'s `max_depth` (64 by
//! default), which is exactly the bound the design doc's "explicit stack, no
//! recursion" note is protecting against — so plain recursion is used; it
//! cannot overflow the stack for any path `SafePath::parse` would have
//! accepted in the first place.

use std::collections::HashSet;
use std::sync::Arc;

use sc_meta::{Aggregate, MetaStore};
use sc_vfs::{FileId, Kind, SafePath, ShareId, ShareRoot};
use parking_lot::Mutex;

impl crate::Core {
    /// Walk from the share root down to `path`, allocating (or reusing) a
    /// fileid for every directory component along the way. Files never need
    /// one here.
    pub(crate) fn ensure_fileid_chain(
        &self,
        root: &ShareRoot,
        share: ShareId,
        path: &SafePath,
    ) -> anyhow::Result<FileId> {
        let max_depth = root.policy().max_depth;
        let mut cur_id = FileId::new(0);
        let mut cur_path = SafePath::root();
        for comp in path.components() {
            // `join_existing`: `path` was resolved before it got here, so this
            // is re-walking something that is on disk, not naming anything new.
            // With `join` the whole chain refused an `a:b` ancestor, and since
            // this is what a permanent delete calls to charge quota back, a
            // folder under one could not be deleted at all.
            cur_path = cur_path.join_existing(comp.as_str(), max_depth)?;
            let st = root.stat(&cur_path)?;
            let is_dir = st.kind == Kind::Dir;
            cur_id = self.meta.fileid(share, cur_id, comp.as_str(), &st, is_dir)?;
        }
        Ok(cur_id)
    }

    /// Public door into [`Self::ensure_fileid_chain`], for a caller that
    /// needs a *stable* id for a single path right now rather than the
    /// aggregate that chain was originally built to compute.
    ///
    /// `sc_core::Entry::id` is populated by a pure lookup (`ops.rs`'s
    /// `build_entry`, via `MetaStore::lookup_fileid`) that never allocates —
    /// "lazy allocation": a plain listing of a share with
    /// a million files must not write a million rows just because someone
    /// browsed it. That is the right default for the native API and the web
    /// UI, which have no protocol reason to *need* an id for an entry they
    /// are merely displaying.
    ///
    /// The compat layer is different: `oc:id`/`oc:fileid`
    /// is the key a real client's local sync journal is built on
    /// (`sc-compat-nc/src/props.rs`'s `NcPropSource::emit` doc comment), so
    /// for *that* caller "no id yet" is not an acceptable answer — leaving
    /// it absent (and, before this existed, falling back to a shared
    /// placeholder) means two different resources report the same id, which
    /// a client reads as "these are the same file". This is the honest
    /// alternative promised instead of that placeholder: a real, persisted
    /// row, allocated cheaply (one `INSERT ... ON CONFLICT` per path
    /// component not already known — see `MetaStore::fileid`), and stable
    /// from that point on because it is a durable rowid, not recomputed
    /// per request.
    ///
    /// No ACL check here, deliberately: like [`Self::aggregate`] (the
    /// existing public caller of the same chain), this exists to give an
    /// *already-resolved and already-permission-checked* entry a persisted
    /// identity, not to answer "may this caller see this path" a second
    /// time — that question was already asked and answered by whatever
    /// `resolve`/`list`/`stat_entry` call produced the entry in the first
    /// place.
    pub fn ensure_fileid(&self, share: ShareId, path: &SafePath) -> anyhow::Result<FileId> {
        let root = self
            .shares
            .get(&share)
            .ok_or_else(|| anyhow::anyhow!("ensure_fileid: unknown share {}", share.get()))?
            .root
            .clone();
        self.ensure_fileid_chain(&root, share, path)
    }

    /// Mark `path`'s ancestor chain (parent up to the share root) dirty.
    /// Ancestors that never got a fileid allocated are simply skipped —
    /// absence from `diretag` already means "must recompute", so there is
    /// nothing to mark ("our own writes").
    /// Best-effort: any I/O error along the way just stops early rather than
    /// failing the caller's operation, which has already committed on disk.
    pub fn mark_dirty(&self, share: ShareId, path: &SafePath) {
        let Some(entry) = self.shares.get(&share) else {
            return;
        };
        let root = entry.root.clone();
        drop(entry);

        let max_depth = root.policy().max_depth;
        let mut ancestors = vec![SafePath::root()];
        let mut cur = SafePath::root();
        for comp in path.parent().components() {
            // Same reason, and this one failed quietly: the `break` stopped
            // marking ancestors dirty part-way, so every directory above an
            // awkward name kept serving a stale aggregate ETag.
            cur = match cur.join_existing(comp.as_str(), max_depth) {
                Ok(p) => p,
                Err(_) => break,
            };
            ancestors.push(cur.clone());
        }

        // The share root itself is never given a real `node` row (it has no
        // parent to be named under — `ensure_fileid_chain` returns the
        // `FileId(0)` sentinel for it without inserting anything), but its
        // aggregate *is* still cached in `diretag` keyed by that sentinel.
        // `lookup_fileid` can never find a row for it (there isn't one), so
        // it has to be pushed unconditionally rather than discovered.
        let mut chain = vec![FileId::new(0)];
        for anc in ancestors.iter().skip(1) {
            let Ok(st) = root.stat(anc) else { continue };
            if let Ok(Some(id)) = self.meta.lookup_fileid(share, &st) {
                chain.push(id);
            }
        }
        if let Err(e) = self.meta.mark_dirty_chain(share, &chain) {
            // Leaving this alone would let every cached aggregate in the
            // chain keep reading as fresh, and a sync client that polls a
            // directory ETag would conclude nothing changed. Falling back to
            // the whole-share bump costs a recompute for directories that did
            // not change and loses nothing.
            tracing::warn!(error = %e, share = share.get(),
                           "dirty-marking failed; invalidating the whole share instead");
            if let Err(e) = self.invalidate_share(share) {
                tracing::error!(error = %e, share = share.get(),
                                "whole-share invalidation also failed; cached directory ETags may be stale");
            }
        }
    }

    /// O(1) whole-share invalidation: bump the share's generation counter
    /// (`sc-meta`'s `bump_share_gen`) so every cached directory aggregate
    /// reads as stale on its next lookup, without walking or naming a
    /// single path. `sc-watch` calls this when its own dirty queue
    /// overflows `full_threshold`, or the OS reports a lost/overflowed
    /// batch of events — in both cases enumerating which paths changed is
    /// no longer possible or worth it.
    pub fn invalidate_share(&self, share: ShareId) -> anyhow::Result<u64> {
        self.meta.bump_share_gen(share)
    }

    /// Recursive directory aggregate ETag, cached in `sc-meta`'s `diretag`
    /// table and single-flighted per directory fileid so concurrent readers
    /// of the same stale directory don't all recompute it at once.
    pub fn aggregate(&self, share: ShareId, vpath: &SafePath) -> anyhow::Result<Aggregate> {
        let root = {
            let entry = self
                .shares
                .get(&share)
                .ok_or_else(|| anyhow::anyhow!("aggregate: unknown share {}", share.get()))?;
            entry.root.clone()
        };
        let gen = self.meta.share_gen(share)?;
        let target_id = self.ensure_fileid_chain(&root, share, vpath)?;
        let mut visited = HashSet::new();
        let mut held = Vec::new();
        self.compute_aggregate(share, &root, vpath, target_id, gen, &mut visited, &mut held)
    }

    // Each parameter is a distinct, differently-typed piece of single-flight
    // walk state (share id, fs root, current path, current fileid,
    // generation stamp, cycle-guard set, deadlock-guard vec) — the compiler
    // already rejects an order mix-up since no two are the same type.
    // Bundling them into a struct would only move the same 7 fields into a
    // literal at this fn's one recursive call site (`aggregate` and self);
    // it wouldn't reduce risk or improve readability.
    #[allow(clippy::too_many_arguments)]
    fn compute_aggregate(
        &self,
        share: ShareId,
        root: &ShareRoot,
        dir_path: &SafePath,
        dir_id: FileId,
        gen: u64,
        visited: &mut HashSet<String>,
        held: &mut Vec<FileId>,
    ) -> anyhow::Result<Aggregate> {
        if let Some(agg) = self.meta.dir_etag(share, dir_id)? {
            return Ok(agg);
        }

        // Single-flight: only one thread computes a given directory's
        // aggregate at a time. Re-check the cache after acquiring the lock —
        // another thread may have just finished.
        //
        // `held` is what keeps this from deadlocking against *itself*. The
        // lock is keyed by fileid, and a fileid is `(share, dev, ino, btime)`
        // — which is unique per directory only as long as the backend reports
        // real inode identity. The portable (Windows dev) backend does not:
        // `dev_ino` returns a constant `(0, 0)`, so two directories created
        // inside the same btime tick collide onto one id. This walk then
        // recurses from a directory into a child carrying the *same* id, and
        // `parking_lot::Mutex` is not reentrant — the thread blocks forever
        // on a lock it is already holding. Observed as a test hanging with
        // no CPU use. Skipping the re-acquire is correct in every case:
        // holding it once already excludes every other thread.
        let lock = if held.contains(&dir_id) {
            None
        } else {
            held.push(dir_id);
            Some(
                self.inflight
                    .entry(dir_id)
                    .or_insert_with(|| Arc::new(Mutex::new(())))
                    .clone(),
            )
        };
        let _guard = lock.as_ref().map(|l| l.lock());

        if let Some(agg) = self.meta.dir_etag(share, dir_id)? {
            return Ok(agg);
        }

        // Cycle detection is keyed on the *path* rather than `(dev, ino)`:
        // this traversal never follows symlinks (only `Kind::Dir` is ever
        // recursed into, and the default `SymlinkPolicy::Deny` means a
        // symlink never reports as `Dir` in the first place), so a plain
        // top-down walk cannot revisit the same directory by a different
        // path — the only theoretical cycle source ('s
        // "hard-linked directory, abnormal FS") is additionally bounded by
        // `SafePath`'s `max_depth`, which `join` enforces on every step
        // below. `(dev, ino)` would be the ideal key, but the portable
        // Windows dev backend reports a constant placeholder for both
        // (`backend/portable.rs`'s `dev_ino`), which makes every directory
        // collide and misfire this check; the path is always meaningful.
        let dir_stat = root.stat(dir_path)?;
        let _ = dir_stat; // still fetched: fails fast if the directory vanished mid-walk
        if !visited.insert(dir_path.to_display_string()) {
            return Ok(Aggregate {
                etag: hex16(blake3::hash(b"").as_bytes()),
                rsize: 0,
                rcount: 0,
            });
        }

        let max_depth = root.policy().max_depth;
        let mut names: Vec<String> = root.read_dir(dir_path)?.into_iter().map(|e| e.name.to_string()).collect();
        names.sort(); // readdir order is unstable; sorting is mandatory (§4.5)

        let mut hasher = blake3::Hasher::new();
        let (mut rsize, mut rcount) = (0u64, 0u64);
        for name in &names {
            let child_path = dir_path.join_existing(name, max_depth)?;
            let st = root.stat(&child_path)?;
            let (child_etag, size, count) = if st.kind == Kind::Dir {
                let child_id = self.meta.fileid(share, dir_id, name, &st, true)?;
                let agg =
                    self.compute_aggregate(share, root, &child_path, child_id, gen, visited, held)?;
                (agg.etag, agg.rsize, agg.rcount)
            } else {
                (MetaStore::file_etag(&st), st.size, 1)
            };
            hasher.update(name.as_bytes());
            hasher.update(child_etag.as_bytes());
            rsize += size;
            rcount += count;
        }
        visited.remove(&dir_path.to_display_string());

        let agg = Aggregate {
            etag: hex16(hasher.finalize().as_bytes()),
            rsize,
            rcount,
        };
        self.meta.put_dir_etag(share, dir_id, &agg, gen)?;
        Ok(agg)
    }
}

fn hex16(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(32);
    for &b in &bytes[..16] {
        out.push(HEX[(b >> 4) as usize] as char);
        out.push(HEX[(b & 0xf) as usize] as char);
    }
    out
}

File: go/internal/core/aggregate.go
//go:build linux

package core

import (
	"context"
	"database/sql"
	"encoding/hex"
	"log/slog"
	"sync"

	"github.com/heavycaffeiner/stowcloud/go/internal/store/cache"
	"github.com/heavycaffeiner/stowcloud/go/internal/vfs"
	"lukechampine.com/blake3"
)

// Aggregate is a directory's cached ETag with the recursive size and count
// that were computed alongside it.
type Aggregate = cache.Aggregate

// aggGuard is the single-flight state for one directory's rollup.
type aggGuard struct {
	mu   sync.Mutex
	done bool
}

// Aggregate returns the recursive rollup for a directory: its ETag, and the
// size and count of everything beneath it. The result is cached in the
// rebuildable half and invalidated by generation or dirty-marking.
//
// The ETag is a hash over the children's identities and tokens, so a change
// of any kind under the directory changes it. The directory id is allocated
// lazily: computing a rollup is the one thing that needs a directory's stable
// id to exist, so that is where it is minted.
func (c *Core) Aggregate(ctx context.Context, share ShareID, p vfs.SafePath) (cache.Aggregate, error) {
	root, ok := c.ShareRoot(share)
	if !ok {
		return cache.Aggregate{}, errf(ErrNotFound, "aggregate an unregistered share")
	}
	gen, err := c.cache.ShareGen(ctx, share)
	if err != nil {
		return cache.Aggregate{}, err
	}
	targetID, err := c.ensureFileIDChain(ctx, root, share, p)
	if err != nil {
		return cache.Aggregate{}, err
	}
	return c.computeAggregate(ctx, root, share, p, targetID, gen, nil, &map[cache.FileID]*aggGuard{})
}

// ensureFileIDChain walks from the share root to p, allocating (or reusing) a
// file id for every directory component along the way. Files never need one
// for a rollup.
func (c *Core) ensureFileIDChain(
	ctx context.Context, root *vfs.ShareRoot, share ShareID, p vfs.SafePath,
) (cache.FileID, error) {
	if p.IsRoot() {
		// The share root has no parent to be named under, so nothing is ever
		// inserted for it; its rollup is still cached, keyed by the sentinel.
		return cache.RootID, nil
	}
	var curID cache.FileID = cache.RootID
	curPath := vfs.RootPath()
	for _, comp := range p.Components() {
		var jerr error
		curPath, jerr = curPath.JoinExisting(comp)
		if jerr != nil {
			return 0, jerr
		}
		st, serr := root.Stat(curPath)
		if serr != nil {
			return 0, mapVFSErr(serr)
		}
		id, uerr := c.upsertDir(ctx, share, curID, comp, st)
		if uerr != nil {
			return 0, uerr
		}
		curID = id
	}
	return curID, nil
}

// upsertDir allocates a stable id for a directory, best-effort: a rollup that
// cannot mint an id is a rollup that cannot be cached, which is a slower
// listing, not a failure.
func (c *Core) upsertDir(ctx context.Context, share ShareID, parent cache.FileID, name string, st vfs.Stat) (cache.FileID, error) {
	var id cache.FileID
	err := c.cache.Write(ctx, func(tx *sql.Tx) error {
		var ierr error
		id, ierr = c.cache.Upsert(ctx, tx, share, parent, name, st)
		return ierr
	})
	return id, err
}

// computeAggregate is the recursive rollup. visited guards against a cycle
// that a hard-linked directory could present, though the walk never follows
// symlinks and the path bound is enforced on every join.
func (c *Core) computeAggregate(
	ctx context.Context,
	root *vfs.ShareRoot,
	share ShareID,
	dirPath vfs.SafePath,
	dirID cache.FileID,
	gen uint64,
	held []cache.FileID,
	guards *map[cache.FileID]*aggGuard,
) (cache.Aggregate, error) {
	if agg, ok, err := c.cache.DirEtag(ctx, share, dirID); err != nil {
		return cache.Aggregate{}, err
	} else if ok {
		return agg, nil
	}

	// Single-flight: only one caller computes a given directory's rollup at a
	// time. The guard is keyed by the file id, and re-entrant calls (a native
	// directory and a hard-linked alias can share one) skip re-acquiring.
	guard, ok := (*guards)[dirID]
	if !ok {
		guard = &aggGuard{}
		(*guards)[dirID] = guard
	}
	acquired := false
	if !containsFileID(held, dirID) {
		held = append(held, dirID)
		guard.mu.Lock()
		acquired = true
		defer func() {
			if acquired {
				guard.mu.Unlock()
			}
		}()
	}
	if agg, ok, err := c.cache.DirEtag(ctx, share, dirID); err != nil {
		return cache.Aggregate{}, err
	} else if ok {
		return agg, nil
	}

	dirStat, err := root.Stat(dirPath)
	if err != nil {
		return cache.Aggregate{}, mapVFSErr(err)
	}
	if !dirStat.Kind.IsDir() {
		return cache.Aggregate{}, errf(ErrNotFound, "aggregate a path that is not a directory")
	}

	entries, err := root.ReadDir(dirPath, vfs.HideReserved)
	if err != nil {
		return cache.Aggregate{}, mapVFSErr(err)
	}
	names := make([]string, 0, len(entries))
	for _, e := range entries {
		names = append(names, e.Name)
	}
	insertionSort(names)

	hasher := blake3.New(32, nil)
	var rsize, rcount uint64
	for _, name := range names {
		childPath, jerr := dirPath.JoinExisting(name)
		if jerr != nil {
			continue
		}
		st, serr := root.Stat(childPath)
		if serr != nil {
			continue
		}
		var childEtag string
		var size, count uint64
		if st.Kind.IsDir() {
			childID, aerr := c.upsertDir(ctx, share, dirID, name, st)
			if aerr != nil {
				return cache.Aggregate{}, aerr
			}
			agg, aerr := c.computeAggregate(ctx, root, share, childPath, childID, gen, held, guards)
			if aerr != nil {
				return cache.Aggregate{}, aerr
			}
			childEtag, size, count = agg.Etag, agg.RSize, agg.RCount
		} else {
			childEtag, _ = FileETag(st)
			size, count = st.Size, 1
		}
		hasher.Write([]byte(name))
		hasher.Write([]byte(childEtag))
		rsize += size
		rcount += count
	}

	agg := cache.Aggregate{
		Etag:   hex.EncodeToString(hasher.Sum(nil)[:16]),
		RSize:  rsize,
		RCount: rcount,
	}
	if err := c.cache.Write(ctx, func(tx *sql.Tx) error {
		return c.cache.PutDirEtag(ctx, tx, share, dirID, agg, gen)
	}); err != nil {
		// A rollup that cannot be stored is a slower next listing, not a
		// failure: the walk already committed nothing, so the caller keeps
		// the value it computed.
		c.logger.Warn("a directory aggregate could not be cached", slog.Any("error", err))
	}
	return agg, nil
}

func containsFileID(held []cache.FileID, id cache.FileID) bool {
	for _, h := range held {
		if h == id {
			return true
		}
	}
	return false
}

func insertionSort(names []string) {
	for i := 1; i < len(names); i++ {
		for j := i; j > 0 && names[j-1] > names[j]; j-- {
			names[j-1], names[j] = names[j], names[j-1]
		}
	}
}

// InvalidateShare is the O(1) whole-share invalidation: bump the generation
// counter so every cached aggregate reads as stale on its next lookup, without
// walking or naming a single path. The watcher calls this when it loses a
// batch of events.
func (c *Core) InvalidateShare(ctx context.Context, share ShareID) error {
	return c.cache.Write(ctx, func(tx *sql.Tx) error {
		_, err := c.cache.BumpShareGen(ctx, tx, share)
		return err
	})
}

// markDirty marks the ancestor chain of path (up to the share root) dirty, so
// their cached aggregates are recomputed. Best-effort: the write already
// committed, so a failure here only costs a stale aggregate that a later
// generation bump or read corrects.
func (c *Core) markDirty(ctx context.Context, share ShareID, p vfs.SafePath) {
	root, ok := c.ShareRoot(share)
	if !ok {
		return
	}
	// The share root's aggregate is cached under the sentinel id that is never
	// a real row, so it is pushed unconditionally.
	chain := []cache.FileID{cache.RootID}
	cur := vfs.RootPath()
	for _, comp := range p.Parent().Components() {
		var jerr error
		cur, jerr = cur.JoinExisting(comp)
		if jerr != nil {
			break
		}
		st, serr := root.Stat(cur)
		if serr != nil {
			continue
		}
		if id, found, lerr := c.cache.Lookup(ctx, share, st); lerr == nil && found {
			chain = append(chain, id)
		}
	}
	if err := c.cache.Write(ctx, func(tx *sql.Tx) error {
		return c.cache.MarkDirty(ctx, tx, share, chain)
	}); err != nil {
		c.logger.Warn("dirty-marking failed; invalidating the whole share instead",
			slog.Uint64("share", uint64(share)), slog.Any("error", err))
		if ierr := c.InvalidateShare(ctx, share); ierr != nil {
			c.logger.Error("whole-share invalidation also failed; cached directory ETags may be stale",
				slog.Uint64("share", uint64(share)), slog.Any("error", ierr))
		}
	}
}
